File: python/network.py
import numpy as np
from layer import MaxPooling2D, Conv2D
from activation import Activation
from optimizer import PassThrough
from loss import MSE
import copy

class Network:
    def __init__(self, layers=None, batch=1, learning_rate=0.02, optimizer=PassThrough(), dtype=np.float32):
        self.layers = layers
        self.batch = batch
        self.learning_rate = learning_rate
        self.input_shape = layers[0].input_shape
        self.configured = False
        self.dtype = dtype

        for i in range(len(self.layers)):
            layer = self.layers[i]
            layer.learning_rate = learning_rate
            layer.batch = batch
            layer.dtype = dtype
            layer.optimizer = copy.deepcopy(optimizer)
        self.last_units = layers[-1].units
        if isinstance(layers[-1], Activation):
            self.last_units = layers[-2].units

    def predict(self, X, batch=1):
        prevLayer = None
        for l in self.layers:
            l.configure(X.shape, "TEST", prevLayer)
            prevLayer = l

        ans = np.zeros((X.shape[0], self.last_units))
        for i in range(0, X.shape[0], self.batch):
            tmp = X[i:i+self.batch, :]
            for layer in self.layers:
                tmpbatch = layer.batch
                layer.batch = batch
                tmp = layer.forward(tmp)
                layer.batch = tmpbatch
            ans[i:i+self.batch, :] = tmp
        return ans

    def train(self, X, label, loss=MSE()):
        if self.configured == False:
            prevLayer = None
            for l in self.layers:
                l.configure(X.shape, "TRAIN", prevLayer)
                prevLayer = l
            self.configured = True

        self.Y = X
        for layer in self.layers:
            self.Y = layer.forward(self.Y)

        err = loss.calc(self.Y, label, self.batch)
        err_delta = loss.partial_derivative(self.Y, label)
        for layer in self.layers[::-1]:
            err_delta = layer.backward(err_delta)

        return err

File: cpp/network.cc
#include "network.h"
#include "loss.h"
#include "layer.h"


Network::Network(int layerNum, Loss* loss, vector<Layer*>* layers) : layerNum(layerNum), layers(layers), loss(loss) {
}

Network::Network(Loss* loss, vector<Layer*>* layers) : layers(layers), loss(loss) {
    this->layerNum = layers->size();
}

Network::~Network() {
    for (int i = 0; i < this->layerNum; i++) {
        delete this->layers->at(i);
    }
};


int Network::configure(int batch, float learning_rate, float v_param, phase_t phase) {
    Layer* prevLayer = nullptr;
    for (int i = 0; i < this->layerNum; i++) {
        this->layers->at(i)->configure(batch, learning_rate, v_param, prevLayer, phase);
        prevLayer = this->layers->at(i);
    }
    loss->configure(batch, prevLayer);
    return 1;
}

void Network::train(vector<float> *data, int* label) {
    for (int j = 0; j < this->layerNum; j++) {
        this->layers->at(j)->forward(data);
        data = &this->layers->at(j)->Y;
    }
    this->loss->partial_derivative(data, label);
    vector<float> *e = &this->loss->D;
    for (int j = this->layerNum-1; j >= 0; j--) {
        this->layers->at(j)->backward(e);
        e = &this->layers->at(j)->E;
    }
    return;
}

float Network::getLossError(int* label) {
    return this->loss->error(&this->layers->back()->Y, label);
}
